crates/codegen/runtime/generator/src/parser/codegen/scanner_definition.rs

```diff
@@ -1,132 +1,216 @@
             }
-
-            ScannerDefinitionNode::ZeroOrMore(node) => {
-                let scanner = node.to_scanner_code();
+            model::Scanner::ZeroOrMore { scanner } => {
+                let scanner = scanner.to_scanner_code();
                 quote! { scan_zero_or_more!(input, #scanner) }
             }
 
-            ScannerDefinitionNode::OneOrMore(node) => {
-                let scanner = node.to_scanner_code();
+            model::Scanner::OneOrMore { scanner } => {
+                let scanner = scanner.to_scanner_code();
                 quote! { scan_one_or_more!(input, #scanner) }
             }
-
-            ScannerDefinitionNode::NoneOf(string) => {
-                let chars = string.chars();
+            model::Scanner::Not { chars } => {
+                let chars = chars.iter();
                 quote! { scan_none_of!(input, #(#chars),*) }
             }
-
-            ScannerDefinitionNode::NotFollowedBy(node, lookahead) => {
+            model::Scanner::TrailingContext {
+                scanner: node,
+                not_followed_by: lookahead,
+            } => {
                 let scanner = node.to_scanner_code();
                 let negative_lookahead_scanner = lookahead.to_scanner_code();
                 quote! { scan_not_followed_by!(input, #scanner, #negative_lookahead_scanner) }
             }
-
-            ScannerDefinitionNode::Sequence(nodes) => {
-                let scanners = nodes
+            model::Scanner::Sequence { scanners } => {
+                let scanners = scanners
                     .iter()
                     .map(|e| e.to_scanner_code())
                     .collect::<Vec<_>>();
                 quote! { scan_sequence!(#(#scanners),*) }
             }
+            model::Scanner::Choice { scanners: nodes } => choice_to_scanner_code(nodes),
 
-            ScannerDefinitionNode::Choice(nodes) => {
-                let mut scanners = vec![];
-                let mut non_literal_scanners = vec![];
-                for node in nodes {
-                    if let ScannerDefinitionNode::Literal(string) = node {
-                        scanners.push(string);
-                    } else {
-                        non_literal_scanners.push(node.to_scanner_code());
-                    }
-                }
-                scanners.sort();
-                let mut scanners = scanners
-                    .iter()
-                    // We want the longest literals first, so we prefer the longest match
-                    .rev()
-                    .map(|string| {
-                        let chars = string.chars();
-                        quote! { scan_chars!(input, #(#chars),*) }
-                    })
-                    .collect::<Vec<_>>();
-                scanners.extend(non_literal_scanners);
-                quote! { scan_choice!(input, #(#scanners),*) }
-            }
-
-            ScannerDefinitionNode::CharRange(from, to) => {
+            model::Scanner::Range {
```

Comment:
Would it be better to use a single terminology here, rather than `from/to` *and* `start/end` ?

Reply:
Agreed, I did limit the diff originally but I'm happy to make it more consistent in the end